### DISCUSSION TITLE
RTK: selectors and type inferring

### DISCUSSION THREAD

**Discussion by dakur** [reduxjs/redux-toolkit#3380] [Q&A] [answered]
Hi again, I need please some more help with typings in RTK, this time when defining selectors. This simple selector which takes advantage of `createSelector()` should work in my opinion:

```ts
import {createSelector} from '@reduxjs/toolkit';

function pages(state: State): Page[]
{
  return state.material.pages;
}

function commentsFromAllPages(): Comment[][]
{
  const out = createSelector(
    [pages],
    (pages) => {
      const comments = pages.map((page) => page.comments);
      return comments;
    },
  );
  return out;
}
```

But it's giving me strange typing error (notice also the type at the line with `const out`):

![image](https://user-images.githubusercontent.com/9862581/232737719-952c6ae2-c0d0-4693-87f0-97d97aac9da8.png)

I've checked for generic arguments or something, but did not find anything.. Can you point me in some direction, please?

Thanks in advance.

**Comment by EskiMojo14** [comment, marked as answer]
createSelector returns an instance of a selector. you should use it like this:
```ts
import { createSelector } from '@reduxjs/toolkit';

function selectPages(state: State): Page[] {
  return state.material.pages;
}

const selectCommentsFromAllPages = createSelector(
  [selectPages],
  (pages) => {
    const comments = pages.map((page) => page.comments);
    return comments;
  },
);
```

**Reply by dakur** [reply]
🤦‍♂️ Oh yes, now I see it, thanks!

**Reply by dakur** [reply]
*not really true, see below*

---

Just to clarify – it's enough to remove the return type from the selector:

```diff
---function commentsFromAllPages(): Comment[][]
+++function commentsFromAllPages()
{
  const out = createSelector(
    [pages],
    (pages) => {
      const comments = pages.map((page) => page.comments);
      return comments;
    },
  );
  return out;
}
```

It will be then inferred correctly:

![image](https://user-images.githubusercontent.com/9862581/232743255-54274a83-3b42-4bc1-9fcb-099790142c0d.png)

**Reply by dakur** [reply]
Sorry, it's not, now I understand it's a "replacement" for the function. 🙂
